src/peer_loop.rs

```diff
@@ -1912,33 +1914,37 @@ impl PeerLoopHandler {
                 own_datetime_utc.format("%Y-%m-%d %H:%M:%S"));
         }
 
-        // There is potential for a race-condition in the peer_map here, as we've previously
-        // counted the number of entries and checked if instance ID was already connected. But
-        // this check could have been invalidated by other tasks so we perform it again
-
-        if global_state
-            .net
-            .peer_map
-            .values()
-            .any(|pi| pi.instance_id() == self.peer_handshake_data.instance_id)
+        // There is potential for a race-condition in the peer_map here, as
+        // we've previously counted the number of entries and checked if
+        // instance ID was already connected. But this check could have been
+        // invalidated by other tasks so we perform it again here under a write
+        // lock.
         {
-            bail!("Attempted to connect to already connected peer. Aborting connection.");
-        }
+            let mut global_state = self.global_state_lock.lock_guard_mut().await;
```

Comment:
The global state is only needed for the peer map. I suggest to bind the peer map explicitly. This reduces repeated statements and communicates the fact that only the peer map is being accessed.

```rust
let peer_map = &mut global_state.net.peer_map;
```